mage/main.go

```diff
@@ -154,7 +155,9 @@ func compile(out string, info *parse.PkgInfo, gofiles []string) error {
 	if err != nil {
 		return errors.WithMessage(err, "can't create mainfile")
 	}
-	defer os.Remove(mainfile)
+	if keep == false {
```

Comment:
can you just make this `if !keep {` ?

Also please add a test (I know it's trivial, but I'm trying to make sure new code has tests, and that way we'll know if we break it by accident later).  Tests should go in the mage subdir, since it already has code to set the binary cache to a temp directory that gets cleaned up automatically.